apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -52,3 +52,31 @@ export default asyncHandler(async (req, res, next) => {
   res.locals.client_fingerprint_id = client_fingerprint_id;
   next();
 });
+
+export async function getClientFingerprintId(req: any, res: any) {
+  const user_session_id = await sqldb.queryOptionalRow(
+    sql.select_user_session_id,
+    { session_id: req.session.id },
+    IdSchema,
+  );
+
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.authn_user.user_id,
+    user_session_id: user_session_id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+  };
+
+  let client_fingerprint_id = await sqldb.queryOptionalRow(
+    sql.select_client_fingerprint,
+    params,
+    IdSchema,
+  );
+
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = await sqldb.queryRow(sql.insert_client_fingerprint, params, IdSchema);
+  }
+
+  return client_fingerprint_id;
+}
```

Comment:
Adding an additional function here was necessary as the assessment instance ID has not been created yet when accessing from `studentAssessmentExam` or `studentAssessmentHomework`. For that reason, we cannot use the default function from the `clientFingerprints` middleware.